pydantic_ai_slim/pydantic_ai/mcp.py

```diff
@@ -50,6 +59,166 @@
+            tool_name = tool_execution_data.get('tool_name')
+            tool_arguments = tool_execution_data.get('arguments', {})
+
+            if tool_name is None:
+                return mcp_types.ErrorData(
+                    code=mcp_types.INVALID_PARAMS, message='Missing tool_name in elicitation request'
+                )
+            bridge_ctx = _create_elicitation_context(tool_name)
+            try:
+                result = await _call_tool_for_elicitation(filtered_toolset, tool_name, tool_arguments, bridge_ctx)
+                return mcp_types.ElicitResult(action='accept', content={'result': str(result)})
+            except ValueError as e:
+                # Tool not found or invalid arguments
+                return mcp_types.ErrorData(code=mcp_types.INVALID_PARAMS, message=f'Tool {tool_name} error: {str(e)}')
+            except TimeoutError:
+                return mcp_types.ErrorData(
+                    code=mcp_types.INTERNAL_ERROR, message=f'Tool {tool_name} execution timed out'
+                )
+            except Exception as e:
+                error_msg = (
+                    f'Tool execution failed for {tool_name}: {str(e)}. Check tool arguments and toolset configuration.'
+                )
+                return mcp_types.ErrorData(code=mcp_types.INTERNAL_ERROR, message=error_msg)
+
+        except json.JSONDecodeError as json_error:
+            return mcp_types.ErrorData(
+                code=mcp_types.INVALID_PARAMS, message=f'Invalid JSON in elicitation request: {str(json_error)}'
+            )
+        except Exception as e:
+            return mcp_types.ErrorData(code=mcp_types.INTERNAL_ERROR, message=f'Elicitation callback failed: {str(e)}')
+
+    return elicitation_callback
+
+
+def create_auto_tool_injection_callback(toolset: AbstractToolset[Any]) -> ProcessToolCallback:
+    """Create a callback that auto-injects available tools into run_python_code calls.
+
+    Args:
+        toolset: The toolset containing tools to be discovered and injected
+
+    Returns:
+        A process tool callback that adds available tools to Python code execution
+    """
+    filtered_toolset = _FilteredToolset(toolset)
+
+    async def auto_inject_tools_callback(
+        ctx: RunContext[Any],
+        call_tool_func: CallToolFunc,
+        tool_name: str,
+        tool_arguments: dict[str, Any],
+    ) -> ToolResult:
+        """Auto-inject tools discovered from toolset into run_python_code calls."""
+        if tool_name == 'run_python_code':
+            available_tools: list[str] = []
+
+            try:
+                discovered_tools = await filtered_toolset.get_tools(ctx=ctx)
+                for original_tool_name, _ in discovered_tools.items():
+                    available_tools.append(original_tool_name)
+            except Exception:
+                pass
+
+            if available_tools:
+                tool_arguments = {**tool_arguments, 'tools': available_tools}
```

Comment:
Can we pass this through metadata instead?

Reply:
Actually, what if we make mcp-run-python use elicitiation to request the available tools, so that we only need to set an `elicitation_callback` and not also a `process_tool_call`?